### ISSUE TITLE
What are the hyperparameters to get an average of 2.02 bits per layer for LLaMA2-7b?

### ISSUE THREAD

**Issue by ylsung** [Vahe1994/AQLM#7]
Hi @Vahe1994, I really like the idea of the paper and thank you for releasing the codes.

I am currently researching the codes, and I would like to know how you obtain an average of `2.02` bits per layer for LLaMA2-7b (which is shown in the paper's Table 1). I tried the hyperparameters (`num_codebooks=1`, `codebook_value_nbits=16`, `in_group_size=8`, `out_group_size=1`, `codebook_value_nbits=16`, `scale_nbits=0`, and the rest of them are the default values) in the README, but I get around `2.50` bit for attention weight and `2.2` bit for MLP weight and the average of them cannot be `2.02`. How do I tune the parameters so that I can get the paper's results? Thank you very much!

**Comment by Vahe1994** [comment]
Hi! Thank you for  taking interest into our paper and for kind words.
To get  2 bits we used 1 codebook with size of  2^15 and group size 8. 
More precisely -  use this configuration to get around the same or better result for llama 2 7B than in the paper:
`data/red_pajama_n=4096_4096_context_length.pth --nsamples=2048 --num_codebooks=1 --nbits_per_codebook=15 --in_group_size=8 --relative_mse_tolerance=0.01 --finetune_lr=1e-4 --finetune_adam_beta1=0.90 --finetune_adam_beta2=0.95 --finetune_keep_best --finetune_relative_mse_tolerance=0.001 --finetune_batch_size=32 --local_batch_size=4 --offload_activations`
Data for calibration you can find in HF. See Readme.md for more details.

The configuration that you mentioned will give you around 2.293 bits for llama-2 7B model(for 13B, and 70B overhead will be less).  This will lead much better results in terms of accuracy, but it is more than 2 bits.

**Comment by ryusaeba** [comment]
Hi @Vahe1994, 
The paper results are really impressive. 
May I have the detailed configurations to generate 3 bit (Table 2) and 4 bit (Table 3) results? 
When compared to GPTQ, the improvement is significant. May I know what's the GPTQ configuration you are using?
Thank you again.

**Comment by Vahe1994** [comment]
Hi @ryusaeba!
For 3 bit results we used 2 codebooks with nbits_per_parameter 12.
`ata/red_pajama_n=4096_4096_context_length.pth --nsamples=1024 --num_codebooks=2 --nbits_per_codebook=12 --in_group_size=8 --relative_mse_tolerance=0.01 --finetune_lr=1e-4 --finetune_adam_beta1=0.90 --finetune_adam_beta2=0.95 --finetune_keep_best --finetune_relative_mse_tolerance=0.001 --finetune_batch_size=32 --local_batch_size=4 --offload_activations  --wandb`

This is not exact hyperparameters, but results should be pretty close, may be slightly better/slightly worse.

For GPTQ we used suggested parameters for 3 bits.
`--wbits 3 --percdamp 1e-2 --nsamples 1024 --permutation_order act_order --round_zero 1 --perchannel `

To get a little bit better results use 2048 nsamples.

**Comment by ylsung** [comment]
@Vahe1994, thank you so much for your pointer. I will try it out!

**Comment by ryusaeba** [comment]
Hi @Vahe1994, 
Thanks for the detailed sharing. Few more follow-up questions about GPTQ here. 
Assuem GPTQ experiments are based on https://github.com/IST-DASLab/gptq/blob/main/llama.py for GPTQ experiments. 
Q1: The default configuration is using -1 for groupsize, so GPTQ results didn't take group into considerations. 
Q2:  The GPTQ use using c4 for calibration.
Thank you again :)

**Comment by Vahe1994** [comment]
Hi @ryusaeba ,
1. Using groups in GPTQ increase  avg_bits per parameter, so we indeed used -1 groupsize to get exact 3, 4 bits. With SpQR(based on GPTQ) we are using groups, but keep average bits to be close to 3, 4 bits(exact avg bits can be found in the paper). 
2. All methods, including GPTQ, is using Red Pajama on LLama-2 family models for a calibration set , for a fair comparison.

**Comment by VityaVitalich** [comment]
Hey! Thank you for your valuable work

For now there is no parameter finetune_relative_mse_tolerance in main and finetune scripts

How could i assess the model with the same parameters to match papers quality?

**Comment by Vahe1994** [comment]
Hey, @VityaVitalich!
Thanks for your interest in the work.
As you may noticed, the both repository and results are constantly evolving.
If you want to reproduce the results from the paper, I suggest using one of the early commits, around February 10.

If you're interested in getting better results, you can try the current version of finetuning 
```
python finetune.py \
  --base_model $MODEL_PATH \
  --quant_model $INPUT_PATH \
  --dataset $DATASET_PATH \
  --nsamples=<TOTAL_SIZE> \
  --val_size=<VAL_SIZE> \
  --lr=1e-5 \
  --adam_beta1=0.90 \
  --adam_beta2=0.999 \
  --epochs=5 \
  --early_stop=3 \
  --batch_size=8 \
  --microbatch_size=4 \
  --save $DATA_PATH \
  --gradient_checkpointing ```
